src/nvim/api/vim.c

```diff
@@ -695,6 +695,56 @@ Dictionary nvim_get_color_map(void)
   return colors;
 }
 
+/// Get the mappings for the corresponding scope and mode
+///
+/// @param mode
+/// @param scope
+Array nvim_get_keymap(String mode, String scope)
+  FUNC_API_SINCE(2)
+{
+  Array mappings = ARRAY_DICT_INIT;
+  mapblock_T *current_maphash;
+
+  // Convert the string mode to the integer mode
+  // that is stored within each mapblock
+  int int_mode = get_map_mode((char_u **)(&(mode.data)), 0);
+  int buffer_local = 0;
+
+  if (strcmp(scope.data, "buffer") == 0) {
+    buffer_local = 1;
+  }
+
+  int i = 0;
+  for (i = 0; i < MAX_MAPHASH; i++) {
+    // Get the correct maphash based on global or local hashes
+    current_maphash = get_maphash(i, scope.data);
+
+    if (current_maphash == NULL) {
+      continue;
+    }
+
+    while (current_maphash) {
+      // Check for correct mode
+      if ((int_mode & current_maphash->m_mode) != 0) {
+        typval_T *dict = xmalloc(sizeof(typval_T));
```

Comment:
You don’t neet to allocate `typval_T` with xmalloc, all functions which deal with `typval_T*` normally expect that you pass something allocated *on the stack* because most of time this is exactly what is being done. But you actually need no typval_T: I would suggest to 

1. Allocate (`dict_T *const dict = dict_alloc();`) and free (`dict_free(dict);`) out of the cycle.
2. Reuse allocated dictionary by using `dict_clear(dict)` in the cycle.
3. Since all functions, but `vim_to_object` expect `dict_T*` you may construct `typval_T` right in the function call: `vim_to_object((typval_T[]) { { .v_type = VAR_DICT, .vval.v_dict = dict } })` if I am not mistaking.

Also I see that you allocate `dict`, but I do not see you freeing it, not to mention unreferencing the dictionary. There should be a memory leak.